Strings to symbols

huwd opened the pull request:
Dependent on: Get :+1: from product before merging!

## What
Tracks down everywhere we'd made a key in the session object a `String`
and changes it to to a symbol

## Why
Switching between symbols and strings was the source of [this bug](https://stackoverflow.com/questions/8189416/why-use-symbols-as-hash-keys-in-ruby), after a [bit of chat on slack ](https://gds.slack.com/archives/C010JE8AMPZ/p1584955486216800?thread_ts=1584954108.200700&cid=C010JE8AMPZ), agreement that we should switch them all over as an after launch thing.

After a bit of reserach turns out symbols have some other benefits:
- [Symbols have access to ruby 2.0 keyword arguments](https://chriszetter.com/blog/2012/11/02/keyword-arguments-in-ruby-2-dot-0/)
- [Symbols are kind of like immutable strings and so are a tad faster](https://web.archive.org/web/20180709094450/http://www.reactive.io/tips/2009/01/11/the-difference-between-ruby-symbols-and-strings)

huwd committed:
Change check answers key to symbol in session

huwd committed:
Change date_of_birth key to symbol in session

huwd committed:
Change name key to symbol in session

huwd committed:
Change check answers controller key to symbol in session

huwd committed:
Change contact_details key to symbol in session

huwd committed:
Change date_of_birth children keys to symbol in session

huwd committed:
Change name children keys to symbols in session

huwd committed:
Change support_address children keys to symbols in session

huwd committed:
Change contact_details children keys to symbols in session

leenagupte commented:
@huwd I tried this out locally, and I got an error for fields that don't have a value.  So maybe use `try` anywhere there's a nested key? e.g. `session[:contact_details].try(:phone_number_calls)`

See: https://apidock.com/rails/Object/try
<img width="1445" alt="Screenshot 2020-03-23 at 11 51 43" src="https://user-images.githubusercontent.com/5793815/77314191-169af080-6cfd-11ea-879b-0420a208d75f.png">

leenagupte commented on spec/controllers/coronavirus_form/name_controller_spec.rb:
```diff
@@ -16,21 +16,21 @@
 
 
-    it "sets session variables" do
+    it "sets session variables as sanatized symbolized keys" do
```
Typo: should be "sanitize".

huwd replied:
... thank you! corrected!

huwd commented:
Ah interesting we're not memoizing everywhere in the same way.
will amend that!

huwd commented:
@leenagupte how does .try work compared to .dig?

```ruby
x = {hello: {world: 1}}  
x[:hello][:world]
> 1
x[:hello].try(:world)
> nil
x[:hello].dig(:world)
> 1
x[:hello].dig(:non_existant_symbol)
> nil
```

huwd committed:
Standardise memoization of nested hashes

Based around the following ideas:

- Symbols better than strings

- .dig() better than direct access (returns nil instead of errors if key doesn't exist!)

- Flow control to check the session object directly (instead of creating similarly variables that exist for a second then disappear

- Required fields to compare to symbols rather than strings

huwd committed:
Learn to spell sanitize

huwd committed:
Correct spacing on contact details form

huwd committed:
Fix support addres controller test

huwd committed:
Fix missing )

huwd committed:
Fix indentation of end

leenagupte reviewed (approved):
👍 💯 Looks good to me!

If you have time, it would be great if you could fix up some of your commits, but it's not a blocker.

bilbof reviewed (dismissed):
From my understanding, sessions [don't have indifferent access](https://www.ruby-forum.com/t/session-with-indifferent-access/54244), so it would be better to use strings wherever possible when interacting with `session`.

leenagupte commented:
> From my understanding, sessions [don't have indifferent access](https://www.ruby-forum.com/t/session-with-indifferent-access/54244), so it would be better to use strings wherever possible when interacting with `session`.

Is that because they're always stored as strings? https://stackoverflow.com/questions/10307412/should-i-use-quotes-or-colons-for-session-keys-in-ruby-on-rails

bilbof commented:
@leenagupte I have a feeling this is the behaviour, but I haven't tested it. I think you can store them using symbols or strings, but when you retrieve them later, the nested objects will be strings. Is that right?

```ruby
# first controller
session[:object] = { symbolised: :value }

# next controller
session[:object] == { "symbolised" => :value }
```

bilbof commented:
@leenagupte sorry, I was wrong there. Using symbols or strings seems to be fine. Maybe I was being inconsistent with the values I was calling them with, which this change would fix :)

```ruby
    session[:object] ||= { symbolised: :value }
    puts session[:object]
    puts session.dig(:object, :symbolised)
    puts session[:object][:symbolised]
```

leenagupte committed:
Fix support_address controller and tests

leenagupte commented:
@huwd I've rebased this against master.

leenagupte committed:
Update AnswerHelper

leenagupte committed:
Update contact details test

leenagupte committed:
Update name test

leenagupte committed:
Update support address test

leenagupte committed:
Fix date of birth controller

leenagupte committed:
Fix nhs number controller

leenagupte committed:
Use dig to find nested params in views.

leenagupte committed:
Use dig to get the name

leenagupte committed:
Stop memoising session in support_address

We're using `dig` to get the session values in the view, so we don't need to do this.

leenagupte committed:
Stop memoising session in date_of_birth

If we use `dig` in the views, we don't need to do this.

huwd commented:
> @huwd I've rebased this against master.

Legend!

brucebolt reviewed (approved):
This looks great :clap:.

I'm guessing we need to do the same for the business volunteering form?  :trollface:
